### PR TITLE
CTCTRADERS-2865 Calculate Departure Status from messages

### PR DESCRIPTION
**Pull request by carrielies**

### CODE REVIEW

File: app/controllers/MessagesController.scala [hunk 1]
```diff
@@ -79,27 +79,23 @@ class MessagesController @Inject()(
                   request.body
                 ) match {
                 case Right(message) =>
-                  StatusTransition.targetStatus(request.departure.status, MessageReceivedEvent.DeclarationCancellationRequest) match {
```
Comment (adamconder) [inline app/controllers/MessagesController.scala:82]: Does removing the call to targetStatus mean we now have different logic for a cancellation request?

Comment (carrielies) [reply app/controllers/MessagesController.scala:82]: The entire DepartureStatus has been removed from the mongodb and is instead calculated from the messages. In the long run we won't even be sent this value

File: app/models/MessageType.scala [hunk 2]
```diff
@@ -62,6 +62,83 @@ object MessageType extends Enumerable.Implicits {
         values.find(_.rootNode == nodeSeq.head.label)
     }
 
+  implicit val ordering: Ordering[MessageType] = (x: MessageType, y: MessageType) => {
```
Comment (drnaylor-oc) [inline app/models/MessageType.scala:65]: I'm kind of nervous about this for a couple of reasons: 

* This is brittle. I don't see any problems with what you've written in that I can see you'll get a consistent ordering every time you use this, but in the very unlikely event that we ever need to add another status (if this gets reused for future phases, I guess?) or we want to rejig the order for some reason, then we have to go through all of this - and it's not guaranteed that anyone will remember to do so or that the written tests will pick it up - there are always edge cases!
* I _think_ this compiles to a set of if/else statements in the bytecode (unfortunately, it can't compile to lookupswitch and tableswitch instructions as they only operates on ints) and if you're ordering any sort of list where, say, `XMLSubmissionNegativeAcknowledgement` is being compared against something, a _lot_ of unnecessary comparisons could be made. I'm not sure this is going to get called all that much, but I'd rather avoid any potential bottlenecks where possible. 

Would it be better if we give each of the `MessageType`s here an integer ordering parameter and just compare on that? The JVM is used to comparing numbers and I suspect that it'll be able to optimise these comparisons far better if we can do that.

File: app/utils/MessageTypeUtils.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    val previousMessage = messagesList.sortBy(_.dateTime).takeRight(2).head
+
+    val messagesWithSameDateTime = messagesList.filter(_.dateTime == previousMessage.dateTime)
+
+    val previousMessageType = if (messagesWithSameDateTime.size == 1) {
+      previousMessage.messageType
+    } else {
+      currentStatus match {
+        case DepartureStatus.DeclarationCancellationRequestNegativeAcknowledgement =>
+          if (previousMessage.messageType == MessageType.DepartureDeclaration | previousMessage.messageType == MessageType.DeclarationCancellationRequest) {
```
Comment (drnaylor-oc) [inline app/utils/MessageTypeUtils.scala:54]: Minor nitpick: use `||` instead of `|`

**Review by drnaylor-oc — COMMENTED** [review]
Couple of thoughts reading through this - nothing exactly breaking but some musings...

File: app/models/MessageType.scala (hunk 2 above)
Comment (carrielies) [reply app/models/MessageType.scala:65]: That makes sense. I didn't actually write this, I moved it from the frontend, but I'll take a closer look at it

Comment (carrielies) [reply app/models/MessageType.scala:66]: We've re-written this. The id word fine apart from one edge case where one status' position in the list is variable. But I've dealt with that in a case statement

Comment (DavidGregory084) [reply app/models/MessageType.scala:66]: Sorry to bikeshed this even more but I think the entire declaration can now be replaced by
```scala
implicit val ordering: Ordering[MessageType] = Ordering.by(_.id)
```

Comment (carrielies) [reply app/models/MessageType.scala:66]: It can't. The problem is XMLSubmissionNegativeAcknowledgement can be returned after both DepartureDeclaration and DeclarationCancellationRequest, which aren't next to each other in the priority order

So if we get either of those status with an XMLSubmission... then we return XMLSubmission...
But if we get XMLSubmission... with any other status, then we return that.

So I placed XMLSubmission... after DepartureDeclaration and had to add a match statement for the one edge case of DeclarationCancellationRequest and XMLSubmission...

File: app/models/DepartureUpdate.scala [hunk 4]
```diff
@@ -28,23 +28,11 @@ sealed trait DepartureUpdate
...
-    case (_: DepartureStatusUpdate, c: CompoundStatusUpdate)  => c
-
-    case (_: MessageStatusUpdate, x: MessageStatusUpdate)   => x
-    case (m: MessageStatusUpdate, a: DepartureStatusUpdate) => CompoundStatusUpdate(a, m)
-    case (_: MessageStatusUpdate, c: CompoundStatusUpdate)  => c
-
-    case (_: CompoundStatusUpdate, x: CompoundStatusUpdate)  => x
-    case (c: CompoundStatusUpdate, a: DepartureStatusUpdate) => c.copy(departureStatusUpdate = a)
-    case (c: CompoundStatusUpdate, m: MessageStatusUpdate)   => c.copy(messageStatusUpdate = m)
+    case (_: MessageStatusUpdate, x: MessageStatusUpdate) => x
```
Comment (DavidGregory084) [inline app/models/DepartureUpdate.scala:31]: Can you explain what's happening with this change? I don't love the `DepartureUpdate` thing at all but this would seem to completely break it at least remove any point in its existence. Can `DepartureUpdate` just be removed now?

File: app/models/MessageType.scala (hunk 2 above)
Comment (DavidGregory084) [reply app/models/MessageType.scala:66]: Oh I see, I couldn't see those special cases in the original diff

File: app/models/DepartureUpdate.scala (hunk 4 above)
Comment (carrielies) [reply app/models/DepartureUpdate.scala:31]: So looking at the old code DepartureUpdate is a sealed trait used to combine updates to the messageStatus and DepartureStatus. We no longer store the DepartureStatus as it's calculated from the messages.

So, yes, I believe that I can just delete DepartureUpdate. I'd missed that. Are you happy with what's happening, and can I go ahead and delete this class (renaming the file to MessageStatusUpdate

Comment (DavidGregory084) [reply app/models/DepartureUpdate.scala:31]: Yep sure that would be great!

Comment (carrielies) [reply app/models/DepartureUpdate.scala]: Ok, I've updated that. Running the tests also updated the copyright to 2022. But that's fine

**Review by DavidGregory084 — APPROVED** [review]
